### DISCUSSION TITLE
UploadFile for multiple files: error handling, timeouts and the file-like API

### DISCUSSION THREAD

**Discussion by ixxie** [fastapi/fastapi#12270] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@files_router.post("/api/files/upload")
async def upload(files: List[UploadFile]):
    logger.debug(files)
    for file in files:
        path = os.path.join(base_path, file.filename)
        try:
            with open(path, "wb") as f:
                while contents := file.file.read(1024 * 1024):
                    f.write(contents)
        except Exception as e:
            logger.error("Failed to upload file")
            logger.exception(e)
        finally:
            file.file.close()

    return {
        "status_code": 201,
        "message": f"File upload successful for {", ".join(map(lambda f: f.filename, files))}",
    }
```

### Description

I'm implementing an endpoint for uploading multiple files to a folder in the backend, and the [example in the docs](https://fastapi.tiangolo.com/tutorial/request-files/#multiple-file-uploads) was helpful, but fell short of explaining some nuance regarding error handling and streaming the file.

Our endpoint currently looks pretty much like in the example code above.

My questions:
1. How do I handle timeouts gracefully, especially when I may be sending large number of files through the endpoint?
2. I'm planning to collect the `successes` and `failures` and set the `status_code` and response object based on this; sounds reasonable?
3. In the `while` loop, I follow a practice recommended in [this SO answer](https://stackoverflow.com/a/70657621) to support large files; however I understand `UploadFile` supports a `SpooledTemporaryFile`. Should I be doing something different here?

### Operating System

Windows

### FastAPI Version

0.111.1

### Pydantic Version

2.9.1

### Python Version

3.12.6

**Comment by jfeaver** [comment, marked as answer]
File uploads are hard. It's likely outside the scope of the FastAPI documentation to have a longer tutorial on them since there are many ways to accomplish them. There are, unfortunately, many pitfalls too.

I've been working on a file upload endpoint recently. I've settled on implementing a quick and fragile file upload (that doesn't allow for timeouts and doesn't do well with large file sizes). It's similar to your example file upload (but only does one upload at a time). Then, I'll also add a second endpoint for a resumable, streaming file upload. I'll use this package to do it (or maybe copy/paste/adapt it): [tuspy-fast-api](https://github.com/edihasaj/tuspy-fast-api). There are other TUS packages out there, like [fastapi-tusd](https://github.com/liviaerxin/fastapi-tusd).

Sending multiple files at once makes things hard. If I were to implement a frontend to access your API then I would probably have the frontend accept multiple files in a form input but I would upload them one at a time. If you can, I'd suggest moving to one file upload at a time, especially if you want to have streaming uploads.

Your second point does sound reasonable but difficult to get right. The communication about what was failing and what was succeeding will take some work to get it right. (this difficulty goes away if you do one upload at a time)

Your third point has a link to a SO article with a lot of good info (I'm still learning 😀). For example, I'm seeing [request.stream()](https://www.starlette.io/requests/#body) for the first time. That might be just the ticket for larger files while still implementing a simple file upload (not using TUS).

**Reply by ixxie** [reply]
Thanks for the insights. 

I reached the same conclusion regarding falling back to one-file upload endpoint, and then having the frontend use that repeatedly. Also avoids having to figure out what kind of status code to return for a partial success / failure.

I didn't even think about streaming files. At what filesize order of magnitude does it make sense to consider that?
